hoodie-spark/src/main/scala/com/uber/hoodie/DefaultSource.scala

```diff
@@ -177,93 +192,107 @@ class DefaultSource extends RelationProvider
-      if (hiveSyncEnabled) {
-        log.info("Syncing to Hive Metastore (URL: " + parameters(HIVE_URL_OPT_KEY) + ")")
-        val fs = FSUtils.getFs(basePath.toString, jsc.hadoopConfiguration)
-        syncHive(basePath, fs, parameters)
+      // Create the dataset if not present (APPEND mode)
+      if (!exists) {
+        val properties = new Properties();
+        properties.put(HoodieTableConfig.HOODIE_TABLE_NAME_PROP_NAME, tblName.get);
+        properties.put(HoodieTableConfig.HOODIE_TABLE_TYPE_PROP_NAME, storageType);
+        properties.put(HoodieTableConfig.HOODIE_ARCHIVELOG_FOLDER_PROP_NAME, "archived");
+        HoodieTableMetaClient.initializePathAsHoodieDataset(sparkContext.hadoopConfiguration, path.get, properties);
       }
-      client.close
-    } else {
-      log.error(s"Upsert failed with ${errorCount} errors :");
-      if (log.isTraceEnabled) {
-        log.trace("Printing out the top 100 errors")
-        writeStatuses.rdd.filter(ws => ws.hasErrors)
-          .take(100)
-          .foreach(ws => {
-            log.trace("Global error :", ws.getGlobalError)
-            if (ws.getErrors.size() > 0) {
-              ws.getErrors.foreach(kt =>
-                log.trace(s"Error for key: ${kt._1}", kt._2))
-            }
-          })
+
+      // Create a HoodieWriteClient & issue the write.
+      val client = DataSourceUtils.createHoodieClient(jsc,
+        schema.toString,
+        path.get,
+        tblName.get,
+        mapAsJavaMap(parameters)
+      )
+      val commitTime = client.startCommit();
+
+      val writeStatuses = DataSourceUtils.doWriteOperation(client, hoodieRecords, commitTime, operation)
+      // Check for errors and commit the write.
+      val errorCount = writeStatuses.rdd.filter(ws => ws.hasErrors).count()
+      if (errorCount == 0) {
+        log.info("No errors. Proceeding to commit the write.");
+        val metaMap = parameters.filter(kv =>
+          kv._1.startsWith(parameters(COMMIT_METADATA_KEYPREFIX_OPT_KEY)))
+        val success = if (metaMap.isEmpty) {
+          client.commit(commitTime, writeStatuses)
+        } else {
+          client.commit(commitTime, writeStatuses,
+            Optional.of(new util.HashMap[String, String](mapAsJavaMap(metaMap))))
+        }
+
+        if (success) {
+          log.info("Commit " + commitTime + " successful!")
+        }
+        else {
+          log.info("Commit " + commitTime + " failed!")
+        }
+
+        val hiveSyncEnabled = parameters.get(HIVE_SYNC_ENABLED_OPT_KEY).map(r => r.toBoolean).getOrElse(false)
+        if (hiveSyncEnabled) {
+          log.info("Syncing to Hive Metastore (URL: " + parameters(HIVE_URL_OPT_KEY) + ")")
+          val fs = FSUtils.getFs(basePath.toString, jsc.hadoopConfiguration)
+          syncHive(basePath, fs, parameters)
+        }
+        client.close
+      } else {
+        log.error(s"Upsert failed with ${errorCount} errors :");
```

Comment:
nit: can you also fix this message to state the actual operation ? :) .. my bad

Reply:
good point, will do